src/fheroes2/dialog/dialog_quickinfo.cpp

```diff
@@ -262,6 +262,34 @@ namespace
         return str;
     }
 
+    std::string showTreeOfKnowledgeInfo( const Maps::Tiles & tile, const bool isVisited )
+    {
+        const MP2::MapObjectType objectType = tile.GetObject( false );
+        std::string str = MP2::StringObject( objectType );
+
+        if ( isVisited ) {
+            const Funds & payment = getTreeOfKnowledgeRequirement( tile );
+            str.append( "\n\n(" );
+            if ( payment.GetValidItemsCount() == 0 ) {
+                str.append( _( "treeOfKnowledge|free" ) );
+            }
+            else {
+                const auto rc = payment.getFirstValidResource();
+                str.append( std::to_string( rc.second ) + " " );
+                str.append( Translation::StringLower( Resource::String( rc.first ) ) );
+            }
+            str += ')';
+
+            const Heroes * hero = Interface::GetFocusHeroes();
+            if ( hero ) {
+                str.append( "\n" );
+                str.append( hero->isObjectTypeVisited( objectType ) ? _( "(already gained)" ) : _( "(not gained)" ) );
```

Comment:
We can not rely only on object type here: if you visit one Tree, then all other Trees will show "(already gained)" even if the level was not gained by that tree.
```suggestion
                str.append( hero->isVisited( tile ) ? _( "(already gained)" ) : _( "(not gained)" ) );
```